drivers/src/pci/config_space.rs

```diff
@@ -81,14 +91,74 @@ impl PciConfigSpace {
     }
 
     // Returns the offset of the given address within this PciConfigSpace.
-    fn header_offset(&self, address: Address) -> Option<u64> {
+    fn config_space_offset(&self, address: Address) -> Option<u64> {
         (address.bits() as u64)
             .checked_sub(Address::bus_address(self.bus_range.start).bits() as u64)
             .map(|a| a << PCIE_ECAM_FN_SHIFT)
     }
 }
 
-/// The configuration space for a PCI bus.
+/// The configuration space for a single PCI function. Used to access the registers within the
+/// configuration space of the function.
+pub struct PciFuncConfigSpace {
+    address: Address,
+    base_ptr: *mut Le32,
+}
+
+impl PciFuncConfigSpace {
+    /// Creates a new function config space for the function at `address` with a config space located
+    /// at `base_ptr`.
+    ///
+    /// # Safety
+    ///
+    /// Caller must guarantee that `base_ptr` points to a valid PCIe function config space of length
+    /// `PCIE_FUNCTION_CONFIG_SIZE` and that `PciFunctionConfigSpace` can read and write to that area
+    /// safely.
+    pub unsafe fn new(address: Address, base_ptr: *mut Le32) -> Self {
+        Self { address, base_ptr }
+    }
+
+    /// Returns the PCI bus address of this function.
+    pub fn address(&self) -> Address {
+        self.address
+    }
+
+    /// Reads a 32-bit value from this function's config space at `offset`.
+    pub fn read_dword(&self, offset: usize) -> Result<u32> {
```

Comment:
Is there a need to do unaligned dword reads? Having the word index passed avoided unwrapping at call sites.

Reply:
We could probably avoid non-dword-sized reads, but we'll eventually need to do byte or word-sized writes. I changed these to use `HeaderWord` for now, but am planning to rework the MMIO access here a bit in preparation for config space emulation.